source/Sashimi.Tests/AzureWebAppHealthCheckActionHandlerFixtures.cs

```diff
@@ -112,5 +119,167 @@ public void WebApp_Is_Not_Found()
+        /// </summary>
+        /// <remarks>
+        /// Calamari Tests operate differently on CI (out of process Calamari.exe) and locally (in-process code execution).
+        /// When in-process, the WebRequest.DefaultWebProxy is used. When out-of-process, the EnvVars are used.
+        /// </remarks>
+        private class ProxySettingsMemento : IDisposable
+        {
+            private readonly List<IDisposable> mementos;
+
+            public ProxySettingsMemento(string hostname, int port)
+            {
+                mementos = new List<IDisposable>();
+
+                SetLocalEnvironmentProxySettings(hostname, port);
+                SetCiEnvironmentProxySettings(hostname, port);
+            }
+
+            private void SetLocalEnvironmentProxySettings(string hostname, int port)
+            {
+                var proxySettings = new UseCustomProxySettings(hostname, port, null!, null!).CreateProxy().Value;
+                mementos.Add(new AutoResetMemento<IWebProxy?>(() => WebRequest.DefaultWebProxy, v => WebRequest.DefaultWebProxy = v, proxySettings));
+            }
+
+            private void SetCiEnvironmentProxySettings(string hostname, int port)
+            {
+                mementos.Add(new EnvironmentVariableMemento(EnvironmentVariables.TentacleProxyHost, hostname));
+                mementos.Add(new EnvironmentVariableMemento(EnvironmentVariables.TentacleProxyPort, $"{port}"));
+            }
+
+            public void Dispose()
+            {
+                mementos.ForEach(m => m.Dispose());
+            }
+        }
+
+        /// <summary>
+        /// Sets an Environment Variable for the lifetime of the EnvironmentVariableMemento object.
+        /// Original setting is restored on dispose.
+        /// </summary>
+        private class EnvironmentVariableMemento : AutoResetMemento<string>
+        {
+            public EnvironmentVariableMemento(string variableName, string testValue)
+                : base(() => Environment.GetEnvironmentVariable(variableName),
+                    value => Environment.SetEnvironmentVariable(variableName, value),
+                    testValue)
+            {
+            }
+        }
+
+        /// <summary>
+        /// A not-quite implementation of the Memento pattern, which enables a single original value to be remembered,
+        /// temporarily overridden with a new value, and then automatically set back to the original value at the end
+        /// of the object's lifetime.
+        ///
+        /// Useful for managing potentially-shared state, though in a simplistic way.
+        /// </summary>
+        /// <remarks>
+        /// Does not attempt to handle concurrency: may produce unexpected results if multiple AutoResetMementos
+        /// are modifying the same shared state (eg Environment variables) across threads.
+        /// </remarks>
+        /// <typeparam name="TValue"></typeparam>
+        private class AutoResetMemento<TValue> : IDisposable
```

Comment:
This looks like an implementation of "setup - test - teardown" pattern, how does this compare with just using `[Setup]` and `[Teardown]` of the test framework?

Reply:
Yeah you're probably right, the code is cute but it's probably unnecessarily complex given the scope of what actually needs to happen here. I'll pare it back.

Reply:
As I went to change it, I realised the reason I went for this in the first place: it allows a clean way, for a single test, to set things up and be sure that they will be restored to their original state, every time.

Setup and Teardown from the test framework apply to all tests in the fixture, so in order to achieve the same effect, I'll need to pull this test out to a separate fixture. I'll still do it... but there was some reasoning behind the implementation apart from just writing clever code ;)

Reply:
Thanks for the background~ Yep I can see why these infrastructures can be useful in enabling parameterised setups. Pulling the tests out in separate fixture may result in some repetitions in code, but that would be a little easier to understand for readers :)

Reply:
Just pushed the changes.